### PR TITLE
ZF3 component compatibility

### PR DESCRIPTION
**Pull request by webimpress**
Work In Progress. Not ready to review yet.

**TODO:**
- [x] Exposed to `zend-component-installer` as modules (`ZF\Apigility\Doctrine\Admin` and `ZF\Apigility\Doctrine\Server`)
- [x] composer scripts
- [x] use `::class` notation everywhere
- [x] `Server\Validator` factories updated to be compatible with ZF3
  - [x] `NoObjectExistsFactory` + tests
  - [x] `ObjectExistsFactory` + tests
- [x] convert all invokables to aliases+factories
- [x] TESTS ! ! !
  - [x] enable `Admin` module tests (never worked, wrong configuration in `phpunit.xml.dist`)
  - [x] tests for extracted Admin module factories:
    - [x] `DoctrineAutodiscoveryModelFactory`
    - [x] `DoctrineMetadataServiceResourceFactory`
    - [x] `DoctrineRestServiceModelFactoryFactory`
    - [x] `DoctrineRestServiceResourceFactory`
    - [x] `DoctrineRpcServiceModelFactoryFactory`
    - [x] `DoctrineRpcServiceResourceFactory`
  - [x] get rid of `Bootstrap.php` and `test/bin/rest-tests`
  - [x] Split Server ODM CRUD tests - one logic test in one test method
  - [x] Split Server ORM CRUD tests - one logic test in one test method
  - ~~Add tests for Server module~~ (out of the scope) 
  - ~~use `mikey179/vfsStream` (if possible?)~~ `modules.config.php.dist` for testing modules instead
  - [x] enable coveralls for the library
- [x] remove big parts of commented out code (`Admin/Model/DoctrineRpcServiceModel` - `createController` method)
- ~~fix travis configuration for hhvm (if possible - mongo?)~~
- [x] test with zf-apigility-admin
- [x] update `composer.json` to tagged latest version of dependencies

**Requires other modules to be merged and tagged:**
- [x] [`zfcampus/zf-apigilty-admin`](https://github.com/zfcampus/zf-apigility-admin) ^1.5
- [x] [`phpro/zf-doctrine-hydration-module`](https://github.com/phpro/zf-doctrine-hydration-module/pull/35) ^3.0
- [x] [`doctrine/DoctrineModule `](https://github.com/doctrine/DoctrineModule) ^1.2
- [x] [`doctrine/DoctrineORMModule `](https://github.com/doctrine/DoctrineORMModule) ^1.1
- [x] [`doctrine/DoctrineMongoODMModule`](https://github.com/doctrine/DoctrineMongoODMModule/pull/168) ^0.11

Anything else? Any suggestions?

### CODE REVIEW

**Comment by weierophinney** [comment]
More items for the TODO:

- [ ] short array notation everywhere
- [ ] set minimum supported PHP version to 5.6
- [ ] Update `ZF\Apigility\Doctrine\Admin\Module`:
  - [ ] Remove direct interface implementations (these will be phased out of ModuleManager and pushed to the appropriate components in the near future)
  - [ ] Remove `getAutoloaderConfig()` implementation
  - [ ] Extract factories defined in `getServiceConfig()` into their own classes
- [ ] update phpcs.xml rules to follow those in zf-apigility and zf-apigility-admin, and run across code base

File: src/Admin/Model/DoctrineMetadataServiceResource.php [hunk 1]
```diff
@@ -1,57 +1,64 @@
...
+     * @var ContainerInterface
+     */
     protected $serviceManager;
 
-    public function setServiceManager(ServiceManager $serviceManager)
+    /**
+     * @param ContainerInterface $serviceManager
+     * @return $this
+     */
+    public function setServiceManager(ContainerInterface $serviceManager)
```
Comment (webimpress) [inline src/Admin/Model/DoctrineMetadataServiceResource.php]: @weierophinney should it be reverted? Is it gonna be BC break?

File: src/Admin/Model/DoctrineRestServiceModel.php [hunk 2]
```diff
@@ -187,10 +191,10 @@ public function getServiceManager()
     /**
      * Set service manager
      *
-     * @param ServiceManager $serviceManager
+     * @param ContainerInterface $serviceManager
      * @return DoctrineRestServiceModel
      */
-    public function setServiceManager(ServiceManager $serviceManager)
+    public function setServiceManager(ContainerInterface $serviceManager)
```
Comment (webimpress) [inline src/Admin/Model/DoctrineRestServiceModel.php]: @weierophinney should it be reverted?

File: src/Admin/Model/DoctrineMetadataServiceResource.php (hunk 1 above)
Comment (weierophinney) [inline src/Admin/Model/DoctrineMetadataServiceResource.php]: Actually, the original is incorrect, too. :frowning: (should have been `setServiceLocator(ServiceLocatorInterface $serviceLocator)`).

Regardless, though, changing the typehint represents a BC break, so revert to the original.

File: src/Admin/Model/DoctrineRestServiceModel.php (hunk 2 above)
Comment (weierophinney) [inline src/Admin/Model/DoctrineRestServiceModel.php]: yes, see above.

Ideally, though, these classes should be refactored to accept the specific dependencies they have via the constructor. I did that in a few places in zf-apigility-admin and one or two other modules while prepping for the v3 releases. The one place I kept it in zf-apigility-admin was in the autodiscovery feature — where it is required for determining if services exist.

**Comment by TomHAnderson** [comment]
@webimpress thank you for picking up this library too.  I'm tuned in.

**Comment by jeandormehl** [comment]
Any movement on this?

**Comment by fagundes** [comment]
Can I help in any way?

**Comment by TomHAnderson** [comment]
@fagundes the problem when I stopped trying was Doctrine was behind ZF2 development so far that we couldn't even use DoctrineModule with ZF 2.6+

It appears this is stuck on (continuing to) wait on Doctrine.

**Comment by fagundes** [comment]
@TomHAnderson I see. Yes, it is stuck.

**Comment by respinoza** [comment]
@webimpress Are we still waiting for the DoctrineModule? They support ZF3 now in dev-master but no idea when they will tag it.

**Comment by webimpress** [comment]
@respinoza Yes, we should wait for tagged version. I know there are still some issues with doctrine modules and maybe this is the reason why it is not tagged yet...

**Comment by shandyDev** [comment]
@webimpress DoctrineModule and DoctrineORMModule released with new tags! DoctrineMongoODMModule ZF3 pull request also was merged.

**Comment by webimpress** [comment]
@zfcampus/doctrine please review and let me know what do you think.
IMHO adding more test for untested parts can be done later on in separate PR. It is already huge PR...

@weierophinney can you enable please coveralls for the library, please?

**Comment by weierophinney** [comment]
> can you enable please coveralls for the library, please?

Done!

**Comment by weierophinney** [comment]
Assigned @jguittard for review; the initial overview I've done looks incredibly solid, though.

**Comment by jguittard** [comment]
OK, will review within tomorrow!

File: config/admin.config.php [hunk 3]
```diff
@@ -1,175 +1,190 @@
...
-            'zf-apigility-doctrine-rpc-service' => array(
+namespace ZF\Apigility\Doctrine\Admin;
+
+use ZF\Apigility\Doctrine\Server;
+
+return [
+    'router' => [
+        'routes' => [
+            'zf-apigility-doctrine-rpc-service' => [
                 'type' => 'segment',
```
Comment (jguittard) [inline config/admin.config.php]: For consistency with other modern ZF OSS, I'd use `Http\Segment::class` while adding `use Zend\Router\Http;`

Comment (webimpress) [inline config/admin.config.php]: We can't do that. Library is compatible with ZF2 and ZF3, so:
ZF2: `Zend\Mvc\Router\Http\Segment`
ZF3: `Zend\Router\Http\Segment`

Comment (jguittard) [inline config/admin.config.php]: My bad, for BC, can't be done indeed!

**Comment by webimpress** [comment]
@jguittard 

> CRUD tests should not be unit tests but behavior tests. I'll submit a dedicated PR, as I've done them many times with Behat.

I kept is as it was, we can change it, but not sure if it should be done now or later on (after release?)

> Running the tests locally with Mongo extensions (legacy and new one) for 5.6 <= PHP <= 7.0.11 is messy and should be at least described / explained in the README, IMHO.

Yeh, agree it's not straight forward, but the problem is with installation Mongo module. It is different on PHP5.6 and PHP7 and maybe also different depends on OS. I'm not sure if can create general instructions working on all OSes.

**Comment by jguittard** [comment]
@webimpress 
As for the behavior tests, I honestly think it can wait :) It'd just be more consistent and easier to test. And stricto sensus, those CRUD tests _are_ functional tests, not unit.

I agree Mongo documentation does not belong to this module. I'll raise an issue on [Doctrine MongoDB project](https://github.com/doctrine/mongodb).

**Comment by weierophinney** [comment]
> > Running the tests locally with Mongo extensions (legacy and new one) for 5.6 <= PHP <= 7.0.11 is messy and should be at least described / explained in the README, IMHO.
> Yeh, agree it's not straight forward, but the problem is with installation Mongo module. It is different on PHP5.6 and PHP7 and maybe also different depends on OS.

We were able to do this with zend-log and zf-oauth2. It's quite tricky, but possible.

My main concern at this point is that we test this out with an Apigility skeleton. I'd like to see:

- Creation of Doctrine-Connected services, both via ORM and ODM
- Verification that the services respond after creation

@jguittard said he'd do that tonight; once he's verified, I think we can merge and tag. I'll then announce its stability with the Apigility 1.4.1 release (which I'm currently sitting on until this is ready, even though this is considered an optional module; makes for a more interesting release announcement!).

**Comment by weierophinney** [comment]
Just to be clear: My take is that we can do BDD changes for a later release. Right now, we can verify existing functionality works as originally tested, release, and then move forward.

Thanks again for this huge patch, @webimpress !

**Comment by webimpress** [comment]
@weierophinney @jguittard I was testing it using https://github.com/webimpress/zf3-apigility-doctrine. Works fine for me, waiting for your confirmation and than we can release ! :) Thanks!

**Comment by jguittard** [comment]
So, after a passionate long night reviewing and testing the PR, here is my report:

- First, I setup a Vagrant linux based virtual machine so that I could recreate a standard environment with everything I needed (LEMP, MongoDB, tools)
- I then setup a composer project with Apigility 1.4 with zend-component-installer:0.3, ran the admin UI, created an API : fine.
- I then setup Doctrine configuration (driver, connection, one DB, 2 entities, nothing fancy)
- I added @webimpress fork as a composer dependency + DoctrineORMModule
- `composer install` went fine, despite the fact that I had to be cautious on where I injected the dependency (ZF\Apigility\Doctrine\Admin **has to** be injected in the development.config.php.dist file, because of the order of merge config and a hard dependency on `ZF\Apigility\Admin`)
- Result is fine, except the order of the loaded modules and other hard dependencies that cause `Zend\ModuleManager\Exception\MissingDependencyModuleException`
- First, I reordered manually, kept going on the Admin UI : the whole process of handling Doctrine based REST services went fine : GG @webimpress ! (although the service list is not refreshed upon service creation, minor pitfall)
- Then I got stuck on this modules order, until @weierophinney told me to bump `zend-component-installer` to 0.4, which I did. This release resolves the dependencies between modules.
- But I started to encounter fatal errors, due to the fact that several modules were implementing common ModuleManager interfaces for config, bootstrap, init, ... and `ComponentInstaller::includeModuleClasses` (new in 0.4) was eventually hitting the module class.
- Problem is : at this time, Composer has not yet dumped the autoloads, so the interfaces declared in the module classes cannot be resolved, causing errors like :
```
Interface 'Zend\ModuleManager\Feature\ConfigProviderInterface' not found in /vagrant/vendor/doctrine/doctrine-module/src/DoctrineModule/Module.php
```
- Giving it many thoughts, I don't see a proper way to handle this, unless removing the interfaces declarations in the module. @weierophinney already spread this advice for future enhancements in ModuleManager. Yet, DoctrineModule and others ZF module providers should do this too, and the sooner, the better.
- Moreover, discouraging / forbidding Module classes to implement interfaces is not, IMHO, a proper advice, at least for OOP considerations (it's already hard to encourage programmation by contracts!)

- *So*... Should a 1.4.1 release be expected tomorrow, it would either require Doctrine folks to remove interfaces declaration from DoctrineORMModule and Phpro\DoctrineHydrationModule to do the same.
- *OR* to come up with a brilliant idea to find  a way to collect interfaces declaration from the module class and manually require them to allow hitting the class for use in [AbstractInjector:: injectAfterDependencies](https://github.com/zendframework/zend-component-installer/blob/0.4.0/src/Injector/AbstractInjector.php#L196)

- As for the BDD tests in a further release, I'll provide an implementation I've been using several times in different API centric projects with Behat 3.

**Comment by jguittard** [comment]
Further checks : service management (i.e. deleting service from Admin UI) is broken, many configuration entries remain in `config/module.config.php`. The service files are deleted, though.

**Comment by webimpress** [comment]
@jguittard I've just checked removing service and this is what I have in `config/module.config.php`:
```php
<?php
return [
    'router' => [
        'routes' => [],
    ],
    'zf-versioning' => [
        'uri' => [],
    ],
    'zf-rest' => [],
    'zf-content-negotiation' => [
        'controllers' => [],
        'accept-whitelist' => [],
        'content-type-whitelist' => [],
    ],
    'zf-hal' => [
        'metadata_map' => [],
    ],
    'zf-apigility' => [
        'doctrine-connected' => [],
    ],
    'doctrine-hydrator' => [],
    'zf-content-validation' => [],
    'input_filter_specs' => [],
];
```

so there are just empty arrays. The same you will see when you create normal REST/RPC service. I think this is general problem (if this is a problem at all - IMHO not).

About dependencies, I've notice it before, and than forgot about it. I created PRs to Doctrine modules to remove these interfaces but these PR were rejected for now. Here is my fix (hack...) to component installer:
https://github.com/zendframework/zend-component-installer/pull/23

Meanwhile I found another problem. It was already reported here:
https://github.com/zendframework/zend-component-installer/pull/22
By default in apigility skeleton we have following module list:

```php
return [
    'Zend\Db',
    'Zend\Filter',
    'Zend\Hydrator',
    'Zend\InputFilter',
    'Zend\Paginator',
    'Zend\Router',
    'Zend\Validator',
    'ZF\Apigility',
    'ZF\Apigility\Documentation',
    'ZF\ApiProblem',
    'ZF\Configuration',
    'ZF\OAuth2',
    'ZF\MvcAuth',
    'ZF\Hal',
    'ZF\ContentNegotiation',
    'ZF\ContentValidation',
    'ZF\Rest',
    'ZF\Rpc',
    'ZF\Versioning',
    'Application',
];
```
So all modules are with one `\`, and all is good. But after creating new API in apigility admin this file is updated and all modules are with extra backslash:
```php
return [
    'Zend\\Db',
    'Zend\\Filter',
    'Zend\\Hydrator',
    'Zend\\InputFilter',
    ...
```

**Comment by jguittard** [comment]
> So all modules are with one \, and all is good. But after creating new API in apigility admin this file is updated and all modules are with extra backslash:

```php
return [
    'Zend\\Db',
    'Zend\\Filter',
    'Zend\\Hydrator',
    'Zend\\InputFilter',
    ...
```
@webimpress This is normal behavior, this is the way the API config file is duplicated to include existing modules and add the newly created one

**Comment by jguittard** [comment]
@webimpress I've re-run composer installation with [your patch](https://github.com/zendframework/zend-component-installer/pull/23) and (at last) it works fine!
So, I propose that this one is used until we can further discuss this with Doctrine team (next week ZendCon is a good opportunity to be convincing!)

**Comment by jguittard** [comment]
I finally got empty arrays upon service deletion... Might have been related to other issues ^^

**Comment by jguittard** [comment]
@weierophinney this PR can be merged once `zend-component-installer` is (patch) fixed as 0.4.1.

**Comment by weierophinney** [comment]
I've merged zendframework/zend-component-installer#24, which fixes the dependency ordering issues, and zendframework/zend-component-installer#25, which fixes the escaped namespace separator issues.

I thus did the following:

- Cloned a version of the Apigility skeleton and applied the patches to include the latest zend-component-installer release.
- Added a repository pointing to a local install of zf-apigility-doctrine that has this patch applied.
- Installed zf-apigility-doctrine.

I was prompted as expected. I then went into the module configuration, and verified that the order in which modules were added was:

- `DoctrineModule`
- `Phpro\DoctrineHydrationModule`
- `ZF\Apigility\Doctrine\Server`

Additionally, per the changes introduced in zend-component-installer 0.4.0, these all appear before the `Application` module itself.

I then put the application in development mode and launched it. The admin came up correctly, with no errors emitted.

As such, this is ready to merge, and we can finally tag!

Thanks, @webimpress, for the fantastic work on this patch, and @jguittard for doing the real-world testing that uncovered those lingering issues with zend-component-installer!

**Pull request merged**